src/palace/manager/reporting/model.py

```diff
@@ -0,0 +1,71 @@
+from __future__ import annotations
+
+from collections.abc import Generator, Iterable, Sequence
+from dataclasses import dataclass, field
+from functools import cached_property
+from typing import Any, Protocol, TypeVar
+
+from sqlalchemy.orm import Session
+from sqlalchemy.sql import Select
+from sqlalchemy.sql.selectable import TextualSelect
+
+TTabularRowData = Sequence[Any]
+TTabularRows = Iterable[TTabularRowData]
+TTabularHeadings = Sequence[str]
+
+
+@dataclass(kw_only=True, frozen=True)
+class TabularQueryDefinition:
+    """A query for generating a report table."""
+
+    key: str
+    title: str
+    statement: Select | TextualSelect = field(repr=False, hash=False)
+
+    @cached_property
+    def headings(self) -> tuple[str, ...]:
+        """Return the headings for the table's rows."""
+        try:
+            headings = tuple(c.name for c in self.statement.c)
+        except AttributeError as e:
+            raise ValueError(
```

Comment:
Minor: Might want to use `PalaceValueError`. I've been trying to use that instead of `ValueError`, so we have a common base class for any Palace exceptions.